### ISSUE TITLE
Implement changes in Core as described in the paper

### ISSUE THREAD

**Issue by jstolarek** [jstolarek/ghc#142]
[#10833](https://ghc.haskell.org/trac/ghc/ticket/10833)

Richard says:

Updating Core will require modifying types/Coercion.hs, which defines the Coercion datatype. In this case, no new constructors need to be added, but it's possible that some ASSERTs need to be changed. It's also conceivable nothing has to happen in that file, except it's probably good to change the comments on NthCo. Then, update coreSyn/CoreLint.hs (which implements the type-checking rules for Core) to accept the new usage of NthCo. (This will, certainly, need a change.)

Then, you'll need to update the solving algorithm to produce the new NthCo forms. It is less clear to me how to do this. Canonicalisation is when other injective forms (that is, datatypes) are decomposed using NthCo. But, it is reasonable to _replace_, say, [G] Maybe a ~ Maybe b with [G] a ~ b during this process. However, if we have a partially injective type function F (injective in its first parameter but not its second), we can't replace [G] F a b ~ F c d with [G] a ~ c, because this is a loss of information. Instead, we want _both_ [G] F a b ~ F c d and [G] a ~ c, as each constraint offers information the other lacks. (Well, the first clearly implies the second, but the second will allow substitution while the first won't.)

The closest bit in the solver to this sort of thing is with superclasses. For example, if we have [G] Ord a, then we want to get both [G] Ord a and [G] Eq a. But we must be careful that canonicalising [G] Ord a spins off [G] Eq a only once. (As we learn more information about constraints, they often get re-canonicalised. For example, if we learn (a ~ Maybe b), then Ord a gets replaced by Ord (Maybe b) and needs to be re-canonicalised.) So, I conjecture we want to do the same with injective type family constraints, spinning off argument equalities only once.

Bottom line: echo what is done with (given) superclass equalities, and I think it will work out fine.

**Comment by jstolarek** [comment]
For now I have not touched anything in `types/Coercion`.hs and `coreSyn/coreLint.hs`.  First I want to implement deriving of constraints in `TcCanonical`.  So, assuming canonicalization is the right place to do this, my reasoning goes as follows.  From [G] `(F a ~ F b)`, where `F` is injective, I want to create [G] `(a ~ b)`.  The entry point is `TcCanonical.canonicalize`.  Since `(F a ~ F b)` is non-canonical we hit the first case and call `canEvNC`.  We are dealing with an equality predicate, so we follow to `canEqNC` and from there to `can_eq_nc'`.  There is a case in that function that looks like this:

``` haskell
-- Decomposable type constructor applications
-- Synonyms and type functions (which are not decomposable)
-- have already been dealt with
can_eq_nc' _flat _rdr_env _envs ev eq_rel
          (TyConApp tc1 tys1) _ (TyConApp tc2 tys2) _
  | isDecomposableTyCon tc1
  , isDecomposableTyCon tc2
  = canDecomposableTyConApp ev eq_rel tc1 tys1 tc2 tys2
```

Obviously this does not apply to type families, since they are not decomposable, but what puzzles me is the comment saying that TFs have already been dealt with.  I mean, where? I don't see any code in the canonicalization that would do that.

Also, should I be concerned with roles in any way? I think not - with TFs they should always be Nominal.

Finally, I'm not yet sure what to do with Wanteds and Deriveds.

**Comment by goldfirere** [comment]
That comment about TFs already having been dealt with is bogus. But the code there is fine because of the `isDecomposableTyCon`s, which rule out type families. (The comment is fixed in `master`, so don't bother updating.)

Look at how classes work here. This is the pattern you wish to emulate. Specifically, note how `canonicalize` on a `CNonCanonical` calls `canClassNC`, but on a `CDictCan`, it calls `canClass`. The `canClassNC` (the `NC` stands for "non-canonical") function, in turn, calls `canClass` but also `emitSuperclasses`. You want to have a new function `canEqNC` that gets calls for `CNonCanonical` equalities, but not for `CTyEqCan`s. The current `canEqNC` function should really be called `canEq`, because it works on both already-canonicalized equalities and non-canonicalized equalities. (The fact that `CTyEqCan`s get passed to a function named `canEqNC` is likely historical accident.) Then, your new `canEqNC` function emits the new TF givens as appropriate.

What I'm worried here is that it will be fragile. For example, `(F a ~ F b)` would work, in this scenario, but `(F a ~ c, c ~ F b)` might not. Because equality constraints are processed so much more heavily than class constraints, the simple superclass story might, in fact, be insufficient here. What I think we might need to avoid creating the TF givens over and over and over is to have some sort of cache lookup. Currently, there is cache lookup for wanteds, but not for givens. Should that change? Maybe.
@simonpj, any thoughts?

**Comment by simonpj** [comment]
No, we should not do this in the canonicaliser.  Canonicalisation always concerns a SINGLE constraint, and reduces it to canonical form.  This fundep stuff concerns interaction between TWO constraints, and should be dealt with in TcInteract.  See addFunDepWork for classes, for example.

So the place you want is interactFunEq. You can see that the derived equalities for built-in things like a+b are already emitted there.  That's your spot!

There is one wrinkle.  A canonical wanted CFunEqCan always looks like (F tys ~ fuv), where fuv is a "flatten unification variable".   Each such CFunEqCan has a distinct fuv.  Moreover, fuvs are treated as untouchable (ie are not unified).

So if you start with (F t1 ~ F t2), you'll get canonical
 g1:   CFunEqCan:  F t1 ~ fuv1
 g2:  CFunEqCan: F t2 ~ fuv2
 g3:  CTyEqCan: fuv1 ~ fuv2

So to see that the two RHSs are the same, we have to "look through" the CTyEqCans.  That is what the calls to lookupFlattenTyVar are doing in interactFunEq.  It's not terribly beautiful, but works fine.

Now, I think the bug you have with #9587 is that we ALREADY have inert g1, g2, and then we are processing a new constraint g3, the CTyEqCan.  Even though g3 can't rewrite g1 or g2, we still want to kick out g1, because when we re-process g1 (with g3 now inert) we'll spot the fundep.  

This is a bad current bug, with (n+m) types, independent of the new injectivity stuff.  It's hard to trigger, though.

In short, in TcSMonad.kickOutRewritable (assuming you are up to date with HEAD), in the line 
    (feqs_out,   feqs_in)   = partitionFunEqs  kick_out_ct funeqmap
you want to have a more specialised kick_out function, one that kicks it out if the RHS of the CFunEqCan is the same as the new_tv.

OK?

**Comment by goldfirere** [comment]
I think the previous comment is more meant for #119 than for this bug. But this seems like it will work for #119.

Do you have any thoughts on dealing with Givens over ITFs?

**Comment by jstolarek** [comment]
I'm not sure which of the two implementation plans outlined above I should follow.  I decided to give Richard's plan a try to see how it works out. Long story short, I have something like this in`TcCanonical`:

``` haskell
canEqNC :: CtEvidence -> EqRel -> Type -> Type -> TcS (StopOrContinue Ct)
canEqNC ev eq_rel ty1 ty2
  = canEq ev eq_rel ty1 ty2 `andWhenContinue` emitInjectiveTFEqualities
```

That mimics the behaviour for type classes, just as Richard suggested. Not surprisingly, the only type of equalities that reach `emitInjectiveTFEqualities` are `CTyEqCan` and, as Simon pointed out:

> So if you start with (F t1 ~ F t2), you'll get canonical
> g1: CFunEqCan: F t1 ~ fuv1
> g2: CFunEqCan: F t2 ~ fuv2
> g3: CTyEqCan: fuv1 ~ fuv2

So the only given equalities I see in `emitInjectiveTFEqualities` are equalities like `[G] cobox_amu :: fsk_amq ~ fsk_ams (CTyEqCan)`. To make progress here I would need to look through `CFunEqCan`s in the same way `lookupFlattenTyVar` looks through `CTyEqCan`s. This would allow to find `g1` and `g2` based on `g3` (referring to Simon's example above). Not sure if that's the right path to follow. Guidance appreciated.

**Comment by jstolarek** [comment]
I did some more hacking, trying to look through `CFunEqCan`s stored in the inert set. I believe that this is not the right way to go: `FunEqMap Ct` that stores the `CFunEqCan`s seems completely unsuitable for the kind of lookup I'm doing. I suppose it can be implemented but it just feels wrong.
